e2e/investigation.md

```diff
@@ -0,0 +1,103 @@
+    dependent on the e2e test that was executed as some tests affect the configuration of the whole dev-env.
+
+It has become clear that there are multiple ways to achieve the parallel test execution
+
+## Possible solutions
+
+**Option 1 (pool of dev-envs)**:
+
+    This solution requires the bootstrapping of separate dev-env instances
+    (both management cluster and target cluster) for individual e2e test cases.
+    (CON: it scales poorly and it would cause problems with locally running tests on a developer's machine)
+
+**Option 2 (solve it in CI)**:
+
+    This approach would require minimal modification to the CAPM3 Makefile and
+    additional some additional logic to the test code to implement the ability to selectively execute e2e test
+    specs (similarly how it is in dev-env integration tests). Option 3 also includes the individually executable tests.
+
+    This approach would basically be implemented by running multiple e2e test jobs in CI that would only test individual e2e
+    test cases. (CON: this would still require multiple hosts in CityCloud and it would be still slow on local dev-envs).
+
+**Option3 (GO approach)**:
+
+Ginkgo test specs can be executed in a parallel fashion as described [here](https://onsi.github.io/ginkgo/#parallel-specs).
+
+    Each individual test case should have it's own cleanup process that reverts the modifications
+    done by the test execution.
+
+    The cleanup process should aim to avoid re-installation of the
+    dev-env. In some cases like pivoting the cleanup process could be implemented as an other
+    test spec/case e.g "re-pivoting".
+
+    Test specs that are executed in parallel mode should be grouped separately from the ones executed sequentially as
+    not all e2e test cases work well together. The tests that are grouped together as parallelly executable would be
+    executed by Ginkgo's native parallel execution feature. The parallelly executed tests would run at the same time
+    in different namespaces on the same developer environment (same management cluster and same managed cluster(s)).
+
+    All the test cases should be able run individually e.g "make test-e2e pivoting update" etc...
+    this could be useful in the future when there will be more test cases.
+
+**NOTE**:
+
+	There is the possibility to combine option 3 and 1 such as in case the modifications is done as part of
+	option 3 would not yield the needed time savings then the test process could be extended with the use of
+	multiple dev-envs and tests could be executed across multiple dev-envs if needed. With this approach
+	the test execution would be still handled by ginkgo parallel test approach and it would require a
+	"big" virtual machine as a build host. This approach also creates a lot of technical depth and increases
+	the complexity of the tests.
+
+## Plan to proceed:
+
+1. All the e2e tests has to be implemented.
+
+2. Option3 has to be implemented anyways to have a maintainable test setup.
```

Comment:
I think you should add a bit more explanation here. The option 3 alone doesn't speed up the current overall test process because we now cannot find any tests that can be in the parallel group. To meet the goal of this study, we need to implement option 3 **with either option 1 or 2**. However, both options has its own cons as you have pointed out in this study. Therefore, I think more discussion or even investigation is needed to overcome the drawback of the options before jumping to the implementation.

Reply:
Thanks for the suggestion I will add more explanation .